### PR TITLE
Beforeinstallprompt (attempt 2)

### PR DESCRIPTION
**Pull request by dmurph**
Apologies for closing the old issue, I can't figure out how to hook this back up to that one. Restarting, and now without a mountain of commits.

Previous feedback is here:
https://github.com/WICG/manifest-incubations/pull/9

***
<a href="https://pr-preview.s3.amazonaws.com/dmurph/manifest-incubations/pull/12.html" title="Last updated on Jul 16, 2021, 7:15 PM UTC (7dfc2c7)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/WICG/manifest-incubations/12/399ab1a...dmurph:7dfc2c7.html" title="Last updated on Jul 16, 2021, 7:15 PM UTC (7dfc2c7)">Diff</a>

### CODE REVIEW

File: index.html [hunk 1]
```diff
@@ -40,7 +40,7 @@
       </p>
     </section>
     <section>
-      <h2 data-dfn-for="">
+      <h2 data-dft-for="">
```
Comment (marcoscaceres) [inline index.html]:
```suggestion
      <h2 data-dfn-for="">
```

File: index.html [hunk 2]
```diff
@@ -129,55 +129,412 @@ <h2 data-dft-for="">
...
+        restrict the availability of this feature to cases where, for instance,
+        there are sufficient <a data-cite="appmanifest#installability-signals">
+          installability signals</a> to warrant <a data-cite=
+          "appmanifest#dfn-installed">installation</a> of the web application.
         </li>
       </ul>
+      <p>
+        In any case, the user agent MUST NOT <a>present an install prompt</a>
+        if the document is not <a data-cite=
+        "appmanifest#dfn-is-installable">installable</a>.
```
Comment (marcoscaceres) [inline index.html:310]: Maybe
```suggestion
        if the document is not [=installed web application|installable=].
```

Note that "is installable" is gone from the other spec... maybe it should be resurrected here?

File: index.html [hunk 3]
```diff
@@ -17,7 +17,7 @@
         github: "WICG/manifest-incubations",
         shortName: "manifest-incubations",
         xref: {
-          specs: ["appmanifest"],
+          specs: ["appmanifest", "dom"],
```
Comment (marcoscaceres) [inline index.html:20]: Note, if you `data-cite="DOM"` below, you don't need it here. Up to you which way you use it.

File: index.html [hunk 4]
```diff
@@ -129,55 +129,412 @@ <h2 data-dft-for="">
...
+        an install prompt</dfn>:
+      </p>
+      <ol>
+        <li>Show some user-agent-specific UI, asking the user whether to
+        proceed with installing the app. See <a data-cite=
+        "appmanifest#installation-sec">privacy and security considerations</a>
+        for recommendations relating to this UI. The <var>result</var> of this
+        choice is either {{AppBannerPromptOutcome|accepted}}
+          <a data-link-for="AppBannerPromptOutcome">accepted</a> or
+          <a data-link-for="AppBannerPromptOutcome">dismissed</a>.
```
Comment (marcoscaceres) [inline index.html]:
```suggestion
          {{AppBannerPromptOutcome\accepted}} or
          {{AppBannerPromptOutcome\dismissed}}.
```

Same globally to the rest of them.

File: index.html [hunk 5]
```diff
@@ -129,55 +129,412 @@ <h2 data-dft-for="">
...
+        "present an install prompt">install prompt being presented</a> or if
+        the <a>steps to install the web application</a> are currently being
+        executed, then abort this step.
+        </li>
+        <li>
+          <a>Queue a task</a> on the <a>application life-cycle task source</a>
+          to do the following:
+          <ol>
+            <li>Let <var>event</var> be a newly constructed
+            <a>BeforeInstallPromptEvent</a> named
```
Comment (marcoscaceres) [inline index.html]: Globally: 
```suggestion
            {{BeforeInstallPromptEvent}} named
```

I lets ReSpec sprinkles the right things for linking.

File: index.html [hunk 6]
```diff
@@ -129,55 +129,412 @@ <h2 data-dft-for="">
...
+            <code>beforeinstallprompt</code>, with its <code>cancelable</code>
```
Comment (marcoscaceres) [inline index.html]:
```suggestion
            `"beforeinstallprompt"`, with its {{Event\cancelable}}
```

File: index.html [hunk 7]
```diff
@@ -129,55 +129,412 @@ <h2 data-dft-for="">
...
+          to the user agent presenting an <a>automated install prompt</a>. It
+          allows the site to cancel the <a>automated install prompt</a>, as
+          well as manually present the <a>site-triggered install prompt</a>.
         </p>
+        <div class="note">
+          If the <a>BeforeInstallPromptEvent</a> is <em>not</em> cancelled, the
+          user agent is allowed to <a>present an install prompt</a>
+          (specifically, an <a>automated install prompt</a>) to the end-user.
+          Canceling the default action (via <a data-cite=
+          "DOM#dom-event-preventdefault">preventDefault</a>) prevents the user
```
Comment (marcoscaceres) [inline index.html]:
```suggestion
          Canceling the default action (via {{Event/preventDefault()}}) prevents the user
```

File: index.html [hunk 8]
```diff
@@ -129,55 +129,412 @@ <h2 data-dft-for="">
...
+          agent from <a>presenting an install prompt</a>. The user agent is
+          free to run <a>steps to notify that an install prompt is
+          available</a> again at a later time.
+        </div>
+        <p data-dfn-for="PromptResponseObject">
+          The <dfn>PromptResponseObject</dfn> contains the result of calling
+          <a data-lt="BeforeInstallPromptEvent.prompt()">prompt()</a>. It
```
Comment (marcoscaceres) [inline index.html]:
```suggestion
          The <dfn>PromptResponseObject</dfn> contains the result of calling
          {{BeforeInstallPromptEvent\prompt()}}. It
```

**Review by marcoscaceres — COMMENTED** [review]
Couple of small fixes + drive-by comments... note, you shouldn't need any `data-link-for=` when linking to IDL things.

File: index.html [hunk 9]
```diff
@@ -129,56 +129,478 @@ <h2 data-dft-for="">
...
+          <abbr title="User Interface">UI</abbr>, directly invoking the steps
+          to <a>present an install prompt</a>.
+        </li>
+        <li>The installation process can occur through a <dfn>site-triggered
+        install prompt</dfn>: the site can programmatically request that the
+        user agent present an install prompt to the user. The user agent MAY
+        restrict the availability of this feature to cases where, for instance,
+        there are sufficient <a data-cite="appmanifest#installability-signals">
+          installability signals</a> to warrant <a data-cite=
+          "appmanifest#dfn-installed">installation</a> of the web application.
```
Comment (marcoscaceres) [inline index.html]:
```suggestion
        there are sufficient signals to warrant [=installed web application|installation=] of the web application.
```

File: index.html [hunk 10]
```diff
@@ -129,56 +129,478 @@ <h2 data-dft-for="">
...
+      </p>
+      <p>
+        The user agent MAY, at any time (only if the document is <a data-cite=
+        "appmanifest#dfn-is-installable">installable</a>), run the <a>steps to
+        notify that an install prompt is available</a> at any time, giving the
+        site the opportunity to show a <a>site-triggered install prompt</a>
+        without the user needing to interact with the user agent UI.
+      </p>
+      <p>
+        To <dfn data-lt=
```
Comment (marcoscaceres) [inline index.html]:
```suggestion
        To <dfn data-local-lt=
```

File: index.html [hunk 11]
```diff
@@ -129,56 +129,478 @@ <h2 data-dft-for="">
...
+          </ol>
+        </li>
+      </ol>
+      <p>
+        The <dfn>steps to notify that an install prompt is available</dfn> are
+        given by the following algorithm:
+      </p>
+      <ol>
+        <li>Wait until the {{Document}} of the <a>top-level browsing
+        context</a> is <a data-cite="html#completely-loaded">completely
```
Comment (marcoscaceres) [inline index.html:346]: "Completely loaded" is not a exported term from HTML, which probably means it's not the right thing to link to?

File: index.html [hunk 12]
```diff
@@ -129,56 +129,478 @@ <h2 data-dft-for="">
...
+        loaded</a>.
+        </li>
+        <li>If there is already an <a data-lt=
+        "present an install prompt">install prompt being presented</a> or if
```
Comment (marcoscaceres) [inline index.html:350]:
```suggestion
        <li>If there is already an <a>install prompt being presented</a> or if
```

File: index.html [hunk 13]
```diff
@@ -129,56 +129,478 @@ <h2 data-dft-for="">
...
+          The <dfn>PromptResponseObject</dfn> contains the result of calling
+          {{BeforeInstallPromptEvent\prompt()}}. It contains one member,
+          <dfn data-link-for="PromptResponseObject">userChoice</dfn>, which
+          states the user's chosen outcome.
         </p>
+        <p>
+          An instance of a <a>BeforeInstallPromptEvent</a> has the following
+          internal slots:
+        </p>
+        <dl>
```
Comment (marcoscaceres) [inline index.html]:
```suggestion
        <dl data-dfn-for="BeforeInstallPromptEvent">
```

File: index.html [hunk 14]
```diff
@@ -129,56 +129,478 @@ <h2 data-dft-for="">
...
+        <section>
+          <h4>
+            <code>prompt()</code> method
+          </h4>
+          <p>
+            The <dfn>prompt</dfn> method, when called, runs the following
+            steps:
+          </p>
+          <ol>
+            <li>If <var>this</var>.<a>[[\userResponsePromise]]</a> is pending:
```
Comment (marcoscaceres) [inline index.html]: Same for any others: 
```suggestion
            <li>If <var>this</var>.{{BeforeInstallPromptEvent/[[userResponsePromise]]}} is pending:
```

File: index.html [hunk 15]
```diff
@@ -129,56 +129,478 @@ <h2 data-dft-for="">
...
+          is available</a> again at a later time.
+        </div>
+        <p data-dfn-for="PromptResponseObject">
+          The <dfn>PromptResponseObject</dfn> contains the result of calling
+          {{BeforeInstallPromptEvent\prompt()}}. It contains one member,
+          <dfn data-link-for="PromptResponseObject">userChoice</dfn>, which
+          states the user's chosen outcome.
         </p>
+        <p>
+          An instance of a <a>BeforeInstallPromptEvent</a> has the following
```
Comment (marcoscaceres) [inline index.html]: s/g
```suggestion
          An instance of a {{BeforeInstallPromptEvent}} has the following
```

File: index.html [hunk 16]
```diff
@@ -129,56 +129,478 @@ <h2 data-dft-for="">
...
+          <p>
+            The <dfn>prompt</dfn> method, when called, runs the following
+            steps:
+          </p>
+          <ol>
+            <li>If <var>this</var>.<a>[[\userResponsePromise]]</a> is pending:
+              <ol>
+                <li>If this event's <a data-cite=
+                "DOM#dom-event-istrusted"><code>isTrusted</code></a> attribute
+                is <code>false</code>, reject
```
Comment (marcoscaceres) [inline index.html]:
```suggestion
                <li>If this event's {{Event/isTrusted}} attribute
                is <code>false</code>, reject
```

File: index.html (hunk 12 above)
Comment (dmurph) [reply index.html:350]: This is linked from a couple sections - how should I resolve links to "present an install prompt"?

**Pull request closed without merging**